---
title: "microarray2"
author: "Seoyeon Oh"
date: "12/2/2021"
output: html_document
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.align = 'center')
knitr::opts_chunk$set(out.width = '60%')
knitr::opts_chunk$set(fig.width=12, fig.height=8)

#library("mogene10sttranscriptcluster.db")
library("ArrayExpress")
library("arrayQualityMetrics")
library("ggplot2")
#library("huex10sttranscriptcluster.db")
library("limma")
library("oligo")
library("siggenes")
library("affy")
#library("pd.huex.1.0.st.v2")
library("RSQLite")
library("DBI")
library("htmltools")
library("biomaRt")
library("tximport")
library("wateRmelon")
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
setwd("/Users/seoyeon/Desktop/MSc Bioinformatics Year 1/Applied High-throughput Analysis/Project/Datasets")
dir <- "/Users/seoyeon/Desktop/MSc Bioinformatics Year 1/Applied High-throughput Analysis/Project/Datasets/Microarray2/"
```

##E-GEOD-64750

### General info

The array used for this dataset is A-AFFY-45 - Affymetrix GeneChip Mouse Genome 430 2.0 [Mouse430_2]. In this experiment, susceptible mice were infected with H5N1 influenza. After 72h RNA was extracted from the lungs of the mice. We used 9 samples of this experiment (susceptible mice).

### Intensity values
Read in the microarray data and examine dimensionsionality of the intensity value matrix.

```{r, message=F}
id <- "E-GEOD-64750"
exonCELs <- list.celfiles(dir)
data.raw_2 <- read.celfiles(paste(rep(dir,length(exonCELs)),exonCELs,sep=""))
```

```{r}
dim(exprs(data.raw_2))
```

This array dataset contains 37 samples, containing 1,004,004 features each.
annotate with sample name and select data

Non-infected control
GSM1579245 1
GSM1579246 1
GSM1579247 1
GSM1579248 1
GSM1579249 1

DBA/2J 72 hours of infection, Highly Pathogenic 
GSM1579250 1
GSM1579251 1
GSM1579252 1
GSM1579253 1

C57BL/6J Non-infected control
GSM1579254
GSM1579255
GSM1579256

Highly Pathogenic, 72 hours
GSM1579257
GSM1579258
GSM1579259
GSM1579260


```{r}
head(exprs(data.raw_2))
```



### Annotation

Here we provide basic sample annotation, including the phenotype of interest and relevant other features (e.g. confounders). This dataset contains array data (A-AFFY-45) of different mice strains (BXD98, BXD97, BXD83, BXD73, BXD68, BXD67 ,BXD43, C57BL/6J, DBA/2J) infected with influenza virus H5N1.

```{r}
sdrf <- read.delim("./E-GEOD-64750.sdrf.txt")
print(sdrf[,c("Source.Name","Comment..Sample_source_name.","Array.Design.REF", "Characteristics..strain.", "Comment..Sample_description.")])
```

###Which samples we are using, and not using:
We will be using samples involving susceptible and resistant mouse strain, DBA/2J (GSM1579245 - GSM1579253) and C57BL/6J (GSM1579254 - GSM1579260) respectively. Each strain was inoculated with H5N1 influenza A virus. We are not using the data from other strains (BXD98, BXD97, BXD83, BXD73, BXD68, BXD67, BXD43) which do not contain non-infected control samples.

```{r}
#Load in the ExpressionFeatureSet object
MouseExp_AE2 <- ArrayExpress("E-GEOD-64750")

#Load in the Affybatch object
MouseExp_pheno2 <- ReadAffy(phenoData=pData(MouseExp_AE2))
```

```{r}
#load first lines of output from the object
head(exprs(MouseExp_pheno2))  
pData(MouseExp_pheno2)  #source name, comment sample description, sample source name, sample title, characteristics, organism, term.source, accession number, sex ...
```

```{r}
filterD <- colnames(data.raw_2)[data.raw_2@phenoData@data$index <= 9]
filterD

filteredD <- data.raw_2[,filterD]
filteredD

dim(exprs(filteredD))  #1004004 features       9 samples
```

```{r}
head(exprs(filteredD))
arrayQualityMetrics(filteredD,outdir="./raw",force=T)
```


```{r}
arrayQualityMetrics(filteredD,outdir="./rawlog",force=T,do.logtransform=T)
```

```{r}
#preprocess the data
#This function converts an AffyBatch object into an ExpressionSet object using the robust multi-array average (RMA) expression measure.
DBA2J_RMA <- oligo::rma(filteredD, background=T)
DBA2J_RMA

arrayQualityMetrics(DBA2J_RMA, force=T)
```


```{r}
## Differential expression analysis with RMA preprocessed data
####################

## Additional preprocessing
samples <- c(replicate(5, "DBA/2J control"), replicate(4, "DBA/2J Highly pathogenic 72hrs"))
pData(DBA2J_RMA)[,2] <- samples
pData(DBA2J_RMA)[,2]
```

```{r}
#head(exprs(mouseRMA))
annotD <- factor(pData(DBA2J_RMA)[,2])
annotD
```


```{r}
designD <- model.matrix(~0+annotD)
colnames(designD)<-c("Control","Infected")
# for strain, 1 represents DBA/2J, 0 represents c57/BL6
designD
```


```{r}
fitD <- lmFit(DBA2J_RMA, designD)
cont.matrix <- makeContrasts(InfectedvsControl="Infected-Control", levels=designD)

fit2D <- contrasts.fit(fitD,cont.matrix) 
fit2D <- eBayes(fit2D)
fit2D

volcanoplot(fit2D)
limma::plotMA(fit2D)
```

```{r}
# DE results
LIMMAoutD <- topTable(fit2D,adjust="BH",number=nrow(exprs(DBA2J_RMA)))
#head(LIMMAout)

## Check intensity values for top results
exprs(DBA2J_RMA)[rownames(exprs(DBA2J_RMA))%in%rownames(head(LIMMAoutD)),]

#mean expression of control
rowMeans(exprs(DBA2J_RMA)[rownames(exprs(DBA2J_RMA))%in%rownames(head(LIMMAoutD)),1:5])

#mean expression of infected
rowMeans(exprs(DBA2J_RMA)[rownames(exprs(DBA2J_RMA))%in%rownames(head(LIMMAoutD)),6:9])
```


```{r}
#Adjustments on p values using Benjamini-Hochberg
LIMMAoutD$diffexpressed <- "NO"
LIMMAoutD$diffexpressed[LIMMAoutD$logFC > 0 & LIMMAoutD$adj.P.Val < 0.05] <- "UP"
LIMMAoutD$diffexpressed[LIMMAoutD$logFC < 0 & LIMMAoutD$adj.P.Val < 0.05] <- "DOWN"

length(which(LIMMAoutD$diffexpressed=="UP"))  # 2699 upregulated genes
length(which(LIMMAoutD$diffexpressed=="DOWN")) # 3398 downregulated genes

#No adjustments on p values
LIMMAoutD$diffexpressed_no_BH <- "NO"
LIMMAoutD$diffexpressed_no_BH[LIMMAoutD$logFC > 0 & LIMMAoutD$P.Value < 0.05] <- "UP"
LIMMAoutD$diffexpressed_no_BH[LIMMAoutD$logFC < 0 & LIMMAoutD$P.Value < 0.05] <- "DOWN"

length(which(LIMMAoutD$diffexpressed_no_BH=="UP"))  # 5164 upregulated genes
length(which(LIMMAoutD$diffexpressed_no_BH=="DOWN")) # 6598 downregulated genes

ggplot(data = LIMMAoutD, aes(x= logFC, y = -log10(adj.P.Val), colour = diffexpressed)) +
 geom_point()+
 theme_bw()+
  geom_hline(yintercept = -log10(0.05), linetype="dashed", color="midnightblue")+
  annotate("text", min(4), 1.3, vjust = -1, label = "Cutoff", color="midnightblue")+
 #ggtitle("Differentiall (unadjusted P-value)") +
 labs(colour = "Differentialy expressed")
 #theme(plot.title = element_text(hjust = 0.5, face = "bold.italic"))

ggplot(data = LIMMAoutD, aes(x= logFC, y = -log10(P.Value), colour = diffexpressed_no_BH)) +
 geom_point()+
 theme_bw()+
  geom_hline(yintercept = -log10(0.05), linetype="dashed", color="midnightblue")+
  annotate("text", min(4), 1.3, vjust = -1, label = "Cutoff", color="midnightblue")+
 #ggtitle("Differentiall (unadjusted P-value)") +
 labs(colour = "Differentialy expressed")

```



```{r}
## Load annotation and sort alphabetically on probe name
annotation_MA <- read.table("A-AFFY-45.adf.txt",header=T,sep="\t",skip=17,fill=T)
print(head(annotation_MA))
annotation_MA <- annotation_MA[sort(annotation_MA$Composite.Element.Name,index.return=T)$ix,]

## Check if all probes are present in both sets
dim(annotation_MA)
dim(LIMMAoutD)

## Double check => "Assumption is the mother of all fuck up's ;)"
sum(annotation_MA$Composite.Element.Name==sort(rownames(LIMMAoutD)))

## Sort LIMMA output alphabetically on probe name
LIMMAout_sorted <- LIMMAoutD[sort(rownames(LIMMAoutD),index.return=T)$ix,]
#LIMMAout_sorted

## Add gene names to LIMMA output
LIMMAout_sorted$gene <- annotation_MA$Composite.Element.Database.Entry.ensembl.
LIMMAout_annot <- LIMMAout_sorted[sort(LIMMAout_sorted$adj.P.Val,index.return=T)$ix,]

#sort by adjusted p value from most significant to least 
LIMMAout_sorted2 <- LIMMAout_sorted[order(LIMMAout_sorted$adj.P.Val, decreasing= F),]

#extract top 50 significant DE genes 
LIMMAout_sorted2[1:50,]$gene

# Have a look at the results and search for other probesets for your DE genes
head(LIMMAout_annot)
LIMMAout_annot[LIMMAout_annot$gene=="ENSMUSG00000045303",]
```

```{r}
filterC <- colnames(data.raw_2)[data.raw_2@phenoData@data$index <= 16 & data.raw_2@phenoData@data$index > 9]
filterC

filteredC <- data.raw_2[,filterC]
```

```{r}
head(exprs(filteredC))
arrayQualityMetrics(filteredC,outdir="./rawC",force=T)
```

```{r}
arrayQualityMetrics(filteredC,outdir="./rawlogC",force=T,do.logtransform=T)
```


```{r}
#preprocess the data
#This function converts an AffyBatch object into an ExpressionSet object using the robust multi-array average (RMA) expression measure.
C57BL6_RMA <- oligo::rma(filteredC, background=T)
C57BL6_RMA

arrayQualityMetrics(C57BL6_RMA, force=TRUE)
```

```{r}
## Differential expression analysis with RMA preprocessed data
####################

## Additional preprocessing
samplesC <- c(replicate(3, "C57/BL6J control"), replicate(4, "C57/BL6J infected"))
pData(C57BL6_RMA)[,2] <- samplesC
pData(C57BL6_RMA)[,2]
```

```{r}
#head(exprs(mouseRMA))
annotC <- factor(pData(C57BL6_RMA)[,2])
annotC
```

```{r}
## Differential expression by LIMMA
# Method as stated in limma package (no intercept, easy for simple model designs)
designC <- model.matrix(~0+annotC)
colnames(designC)<-c("Control","Infected")
designC
```

```{r}
fitC <- lmFit(C57BL6_RMA, designC)
cont.matrixC <- makeContrasts(InfectedvsControl="Infected-Control", levels=designC)

fit2C <- contrasts.fit(fitC,cont.matrixC) 
fit2C <- eBayes(fit2C)
fit2C

volcanoplot(fit2C)
limma::plotMA(fit2C)
```
```{r}
# DE results
LIMMAoutC <- topTable(fit2C,adjust="BH",number=nrow(exprs(C57BL6_RMA)))
#head(LIMMAout)

## Check intensity values for top results
exprs(C57BL6_RMA)[rownames(exprs(C57BL6_RMA))%in%rownames(head(LIMMAoutC)),]

#mean expression of control
rowMeans(exprs(C57BL6_RMA)[rownames(exprs(C57BL6_RMA))%in%rownames(head(LIMMAoutC)),1:3])

#mean expression of infected
rowMeans(exprs(C57BL6_RMA)[rownames(exprs(C57BL6_RMA))%in%rownames(head(LIMMAoutC)),4:7])
```

```{r}
## Load annotation and sort alphabetically on probe name
annotation_MA <- read.table("A-AFFY-45.adf.txt",header=T,sep="\t",skip=17,fill=T)
print(head(annotation_MA))
annotation_MA <- annotation_MA[sort(annotation_MA$Composite.Element.Name,index.return=T)$ix,]

## Check if all probes are present in both sets
dim(annotation_MA)
dim(LIMMAoutC)

## Double check => "Assumption is the mother of all fuck up's ;)"
sum(annotation_MA$Composite.Element.Name==sort(rownames(LIMMAoutC)))

## Sort LIMMA output alphabetically on probe name
LIMMAout_sortedC <- LIMMAoutC[sort(rownames(LIMMAoutC),index.return=T)$ix,]
#LIMMAout_sorted

## Add gene names to LIMMA output
LIMMAout_sortedC$gene <- annotation_MA$Composite.Element.Database.Entry.ensembl.
LIMMAout_annotC <- LIMMAout_sortedC[sort(LIMMAout_sortedC$adj.P.Val,index.return=T)$ix,]

#sort by adjusted p value from most significant to least 
LIMMAout_sorted2C <- LIMMAout_sortedC[order(LIMMAout_sortedC$adj.P.Val, decreasing= F),]

#extract top 50 significant DE genes 
LIMMAout_sorted2C[1:50,]$gene

# Have a look at the results and search for other probesets for your DE genes
head(LIMMAout_annotC)
LIMMAout_annotC[LIMMAout_annot$gene=="ENSMUSG00000045303",]
```

```{r}
par(mfrow=c(1,2))
volcanoplot(fit2D)
volcanoplot(fit2C)
```
```{r}
#Adjustments on p values using Benjamini-Hochberg
LIMMAoutC$diffexpressed <- "NO"
LIMMAoutC$diffexpressed[LIMMAoutC$logFC > 0 & LIMMAoutC$adj.P.Val < 0.05] <- "UP"
LIMMAoutC$diffexpressed[LIMMAoutC$logFC < 0 & LIMMAoutC$adj.P.Val < 0.05] <- "DOWN"

length(which(LIMMAoutC$diffexpressed=="UP"))  # 1401 upregulated genes
length(which(LIMMAoutC$diffexpressed=="DOWN")) # 1031 downregulated genes

#No adjustments on pvalues
LIMMAoutC$diffexpressed_no_BH <- "NO"
LIMMAoutC$diffexpressed_no_BH[LIMMAoutC$logFC > 0 & LIMMAoutC$P.Value < 0.05] <- "UP"
LIMMAoutC$diffexpressed_no_BH[LIMMAoutC$logFC < 0 & LIMMAoutC$P.Value < 0.05] <- "DOWN"

length(which(LIMMAoutC$diffexpressed_no_BH=="UP"))  # 3612 upregulated genes
length(which(LIMMAoutC$diffexpressed_no_BH=="DOWN")) # 4429 downregulated genes

C57_down <- LIMMAoutC[LIMMAoutC$diffexpressed_no_BH=="DOWN", ]
C57_up <- LIMMAoutC[LIMMAoutC$diffexpressed_no_BH=="UP", ]

#a <- subset(LIMMAoutC, LIMMAoutC$diffexpressed_no_BH=="DOWN")

ggplot(data = LIMMAoutC, aes(x= logFC, y = -log10(adj.P.Val), colour = diffexpressed)) +
 geom_point()+
 theme_bw()+
  geom_hline(yintercept = -log10(0.05), linetype="dashed", color="midnightblue")+
  annotate("text", min(4), 1.3, vjust = -1, label = "Cutoff", color="midnightblue")+
 #ggtitle("Differentiall (unadjusted P-value)") +
 labs(colour = "Differentialy expressed")
 #theme(plot.title = element_text(hjust = 0.5, face = "bold.italic"))

ggplot(data = LIMMAoutC, aes(x= logFC, y = -log10(P.Value), colour = diffexpressed_no_BH)) +
 geom_point()+
 theme_bw()+
  geom_hline(yintercept = -log10(0.05), linetype="dashed", color="midnightblue")+
  annotate("text", min(4), 1.3, vjust = -1, label = "Cutoff", color="midnightblue")+
 #ggtitle("Differentiall (unadjusted P-value)") +
 labs(colour = "Differentialy expressed")

```
```{r}
#comparison between the two strains
par(mfrow=c(1,2))
#strain DB2J
ggplot(data = LIMMAoutD, aes(x= logFC, y = -log10(P.Value), colour = diffexpressed_no_BH)) +
 geom_point()+
 theme_bw()+
  geom_hline(yintercept = -log10(0.05), linetype="dashed", color="midnightblue")+
  annotate("text", min(4), 1.3, vjust = -1, label = "Cutoff", color="midnightblue")+
  ylim(-0.5,15)+
 #ggtitle("Differentiall (unadjusted P-value)") +
 labs(colour = "Differentialy expressed")

ggplot(data = LIMMAoutC, aes(x= logFC, y = -log10(P.Value), colour = diffexpressed_no_BH)) +
 geom_point()+
 theme_bw()+
  geom_hline(yintercept = -log10(0.05), linetype="dashed", color="midnightblue")+
  annotate("text", min(4), 1.3, vjust = -1, label = "Cutoff", color="midnightblue")+
  ylim(-0.5,15)+
 #ggtitle("Differentiall (unadjusted P-value)") +
 labs(colour = "Differentialy expressed")
```


```{r}
library(ggvenn)
#extract DE genes
DEgenes_D <- LIMMAout_sorted2$gene[LIMMAout_sorted2$adj.P.Val <= 0.05]
DEgenes_D <- DEgenes_D[DEgenes_D != ""]
#head(DEgenes_D, 10)

DEgenes_C <- LIMMAout_sorted2C$gene[LIMMAout_sorted2$adj.P.Val <= 0.05]
DEgenes_C <- DEgenes_C[DEgenes_C != ""]
#head(DEgenes_C, 10)

overlap <-list('DBA/2J'= DEgenes_D,'C57/BL6'=DEgenes_C)

#Create venn diagram and display all the sets
ggvenn(overlap)
length(DEgenes_C)
length(DEgenes_D)
```






